File: 800/b-taxi.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	// your code goes here
	
	
	    int n;
	    cin>>n;
	    vector<int>arr(n);
	    unordered_map<int,int>map;
	    for(int i=0; i<n; i++)
	    {
	        cin>>arr[i];
	        map[arr[i]]++;
	    }
	    int count=0;
	    if(map.find(4)!=map.end())
	    {
	        
	        count+=map[4];
	       // cout<<"4: "<<count<<endl;

	    }
	    if(map.find(3)!=map.end())
	    {
	        if(map.find(1)!=map.end())
	        {
	         int minpair=min(map[1],map[3]);
	         count+=minpair;
	       //   cout<<"31: "<<count<<endl;

	         map[1]-=minpair;
	         map[3]-=minpair;
	         
	            
	        }
	        if(map.find(3)!=map.end())
	        {
	            count+=map[3];
	           // cout<<"3: "<<count<<endl;
	            map[3]=0;
	        }
	    }
	    if(map.find(2)!=map.end())
	    {
	        if(map[2]%2==0)
	        {
	            count+=map[2]/2;
	           // cout<<"2 even: "<<count<<endl;

	            map[2]=0;
	        }
	        else
	        {
	            count+=map[2]/2;
	            
	           // cout<<"2 odd but pending one: "<<count<<endl;
	            if(map.find(1)!=map.end() && map[1]!=0)
	            {
	               if(map[1]>=2)
	               {
	                   map[1]-=2;
	               }
	               else
	               {
	                   map[1]--;
	               }
	               //cout<<"2 1 1 or 2 1: "<<count<<endl;
	               count++;

	               
	            }else
	            {
	               // cout<<"2 single: "<<count<<endl;
	                count++;
	            }
	        }
	    }
	    if(map.find(1)!=map.end())
	    {
	        if(map.find(1)!=map.end())
	        {
	           if(map[1]>=4)
	           {
	               int four=map[1]/4;
	               //cout<<"4 1: "<<count<<endl;
	               count+=four;
	               map[1]-=four*4;
	           }
	           if(map[1]!=0 && map[1]<4)
	           {
	               //cout<<"1: "<<count<<endl;
	               count++;
	           }
	           
	            
	            
	            	          //  cout<<"1: "<<count<<endl;

	        }
	    }
	    cout<<count<<endl;
	   
	

}

File: 800/b-taxi.py
n=int(input())
arr=list(map(int,input().split()))
mp={}
for i in range(n):
    mp[arr[i]]=mp.get(arr[i],0)+1
# print(mp)
    
# print(arr)
count=0
if 4 in mp:
    count=count+mp[4]
if 3 in mp:
    if 1 in mp:
        mini=min(mp[1],mp[3])
        count=count+mini
        mp[1]=mp[1]-mini
        mp[3]=mp[3]-mini
    if 3 in mp and mp.get(3,0)>0:
        count=count+mp[3]
        mp[3]=0
if 2 in mp:
    if mp[2]%2==0:
        count=count+mp[2]//2
        mp[2]=0
    else:
        count=count+mp[2]//2
        if 1 in mp and mp[1]>0:
            if mp[1]>=2:
                mp[1]=mp[1]-2
            else:
                mp[1]=mp[1]-1
            count=count+1
        else:
            count=count+1
            
if 1 in mp:
    if mp[1]>=4:
        four=mp[1]//4
        count=count+four
        mp[1]=mp[1]-4*four
    if mp[1]!=0 and mp[1]<4:
        count=count+1
        
print(count)   